src/boot/AppDataFetcher.js

```diff
@@ -1,25 +1,47 @@
 /* @flow strict-local */
 
-import React from 'react';
+import { useCallback, useEffect } from 'react';
 import type { Node } from 'react';
 
-import { useSelector, useDispatch } from '../react-redux';
-import { getSession } from '../directSelectors';
-import { doInitialFetch } from '../actions';
+import { useGlobalSelector, useDispatch } from '../react-redux';
+import { getHasAuth } from '../selectors';
+import { registerAndStartPolling } from '../events/eventActions';
+import { sendOutbox } from '../outbox/outboxActions';
+import { initNotifications } from '../notification/notifTokens';
 
 type Props = $ReadOnly<{|
   children: Node,
 |}>;
 
 export default function AppDataFetcher(props: Props): Node {
-  const needsInitialFetch = useSelector(state => getSession(state).needsInitialFetch);
   const dispatch = useDispatch();
 
-  React.useEffect(() => {
-    if (needsInitialFetch) {
-      dispatch(doInitialFetch());
+  const hasAuth = useGlobalSelector(getHasAuth);
+
+  const init = useCallback(async () => {
+    // Init right away if there's an active, logged-in account.
+    // NB `getInitialRouteInfo` depends intimately on this behavior.
+    if (hasAuth) {
+      await dispatch(registerAndStartPolling());
+
+      // TODO(#3881): Lots of issues with outbox sending
+      dispatch(sendOutbox());
+
+      dispatch(initNotifications());
     }
-  }, [needsInitialFetch, dispatch]);
+  }, [dispatch, hasAuth]);
+
+  useEffect(
+    () => {
+      init();
+    },
+    // We just want to init once, just after the store rehydrates. This
+    // component will have begun its life at that time, since it's under
+    // HideIfNotHydrated.
+    //
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [],
```

Comment:
I think this still calls for an explanation of why it's OK that we only call this once, even if `hasAuth` later changes. Along the lines of here:
https://github.com/zulip/zulip-mobile/pull/5300#discussion_r829577313